aidants_connect_web/admin.py

```diff
@@ -168,6 +171,30 @@ class Meta:
         model = Organisation
 
 
+class OrganisationRegionFilter(SimpleListFilter):
+    title = "Région"
+
+    parameter_name = "region"
+
+    def lookups(self, request, model_admin):
+        return [(r.id, r.name) for r in DatavizRegion.objects.all()]
+
+    def queryset(self, request, queryset):
+        region_id = self.value()
+
+        if not region_id:
+            return
+
+        region = DatavizRegion.objects.get(id=region_id)
+        d2r = DatavizDepartmentsToRegion.objects.filter(region=region)
+        first_d2r = d2r.first()
+        q = Q(zipcode__startswith=first_d2r.department.zipcode)
+        for d in d2r.all():
+            q = q | Q(zipcode__startswith=d.department.zipcode)
```

Comment:
Je propose plutôt :

```python
iterable = iter(DatavizDepartmentsToRegion.objects.filter(region=region))
initial = Q(zipcode__startswith=next(it).department.zipcode)
functools.reduce(
    lambda acc, el: acc | Q(zipcode__startswith=el.department.zipcode),
    iterable,
    initial,
)
```

qui a les avantages de ne pas répéter le premier élément et d'être plus concis mais peut-être l'inconvéniant d'être plus cryptique.

Reply:
(J'aime le fonctionnel. C'est beau, le fonctionnel :heart_eyes:)

Reply:
J'avoue qu'à première vue c'est trop cryptique à mon goût et franchement je n'irais pas dire que c'est "beau". :D 
Ça répond toutefois à ma question d'hier quand je demandais comment faire "pythonesquement". Je vais lire la doc de `reduce` pour digérer le truc et voir si je veux bien remplacer ma petite boucle par un petit reduce.

Reply:
Oui, c'est vraiment une question de goût, et de familiarité avec les patterns fonctionnels. `reduce`, c'est plutôt un classique des langages fonctionnels. C'est pour ça que je conçois que tout le monde puisse ne pas être familier avec. Si tu préfère garder la boucle, par exemple pour faciliter la maintenance par d'autres personnes, ça me va aussi.

Reply:
Dans le cœur de Django ils utilisent bien reduce sur les Q, mais avec l'operator "or_", and I think it's beautiful : 

```
reduce(operator.or_, or_queries)
```

Je vais tenter comme ça !

Reply:
Oui, mais tu peux pas parce les éléments sur lesquels tu itères sont pas des `Q`.

Reply:
J'ai fait une compréhension sur mes éléments pour les transformer en groupe de Q, puis je reduce sur mon groupe de Q. Regarde comme c'est Pythonesque maintenant ❤️  : 

https://github.com/betagouv/Aidants_Connect/pull/420/files#diff-155ee91a5a0bd833d19224e81dfcfa5834645bf160d06736690af8d81887d362R197-R200

Reply:
:heart_eyes: